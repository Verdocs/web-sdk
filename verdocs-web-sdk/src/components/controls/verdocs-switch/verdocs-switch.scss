@import '../../../theme.scss';

verdocs-switch {
  font-family: $primary-font;

  .switch {
    display: inline-flex;
    height: 24px;
    width: 44px;
    flex-shrink: 0;
    align-items: center;
    cursor: pointer;
    border-radius: 12px;
    border: transparent;
    transition-property: color, background-color, border-color, text-decoration-color, fill, stroke;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
    transition-duration: 150ms;
  }

  .switch.primary[data-state='checked'] {
    background: $primary-color;
  }

  .switch.primary[data-state='unchecked'] {
    background: $extra-light-border-color;
  }

  .switch.secondary[data-state='checked'] {
    background: $secondary-color;
  }

  .switch.secondary[data-state='unchecked'] {
    background: $extra-light-border-color;
  }

  .slider {
    display: block;
    width: 20px;
    height: 20px;
    border-radius: 20px;
    background: #ffffff;
    box-shadow:
      0 10px 15px -3px rgb(0 0 0 / 0.1),
      0 4px 6px -4px rgb(0 0 0 / 0.1);
    transition-property: transform;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
    transition-duration: 150ms;
  }

  .slider[data-state='checked'] {
    transform: translateX(22px);
  }

  .slider[data-state='unchecked'] {
    transform: translateX(-2px);
  }

  .switch.disabled.primary,
  .switch.disabled.secondary {
    cursor: not-allowed;
    background: $dark-disabled-color;
  }
}
