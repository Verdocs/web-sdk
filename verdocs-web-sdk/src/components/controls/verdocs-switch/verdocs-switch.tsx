import {Component, Prop, h, EventEmitter, Event} from '@stencil/core';

/**
 * Displays a toggle switch
 *
 * ```ts
 * <verdocs-switch checked={sendReminders} onCheckedChange={setSendReminders} />
 * ```
 */
@Component({
  tag: 'verdocs-switch',
  styleUrl: 'verdocs-switch.scss',
})
export class VerdocsSwitch {
  @Prop({mutable: true}) checked: boolean = false;

  /**
   * Select purple or green treatments.
   */
  @Prop() theme: 'primary' | 'secondary' = 'primary';

  /**
   * Should the field be disabled?
   */
  @Prop() disabled: boolean = false;

  @Event({composed: true}) checkedChange: EventEmitter<boolean>;

  render() {
    return (
      <button
        value="on"
        type="button"
        role="switch"
        class={`switch ${this.theme} ${this.disabled ? 'disabled' : ''}`}
        data-state={this.checked ? 'checked' : 'unchecked'}
        aria-checked={this.checked ? 'checked' : 'unchecked'}
        onClick={() => {
          if (!this.disabled) {
            this.checked = !this.checked;
            this.checkedChange?.emit(this.checked);
          }
        }}
      >
        <span data-state={this.checked ? 'checked' : 'unchecked'} class="slider"></span>
      </button>
    );
  }
}
