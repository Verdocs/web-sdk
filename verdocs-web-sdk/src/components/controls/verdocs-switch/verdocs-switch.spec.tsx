import { newSpecPage } from '@stencil/core/testing';
import { VerdocsSwitch } from './verdocs-switch';

describe('verdocs-switch', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsSwitch],
      html: `<verdocs-switch></verdocs-switch>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders as checked and disabled', async () => {
    const page = await newSpecPage({
      components: [VerdocsSwitch],
      html: `<verdocs-switch checked="true" disabled="true"></verdocs-switch>`,
    });
    expect(page.root).toMatchSnapshot();
    const button = page.root.querySelector('button');
    expect(button?.getAttribute('data-state')).toBe('checked');
    expect(button?.classList.contains('disabled')).toBe(true);
  });

  it('emits checkedChange event on click', async () => {
    const page = await newSpecPage({
      components: [VerdocsSwitch],
      html: `<verdocs-switch></verdocs-switch>`,
    });
    const spy = jest.fn();
    page.root.addEventListener('checkedChange', spy);

    const button = page.root.querySelector('button');
    button?.dispatchEvent(new Event('click', { bubbles: true }));
    expect(spy).toHaveBeenCalled();
  });
});
