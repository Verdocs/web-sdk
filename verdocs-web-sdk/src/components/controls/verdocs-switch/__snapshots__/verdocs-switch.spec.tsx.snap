// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-switch renders as checked and disabled 1`] = `
<verdocs-switch checked="true" disabled="true">
  <button aria-checked="checked" class="disabled primary switch" data-state="checked" role="switch" type="button" value="on">
    <span class="slider" data-state="checked"></span>
  </button>
</verdocs-switch>
`;

exports[`verdocs-switch renders with default props 1`] = `
<verdocs-switch>
  <button aria-checked="unchecked" class="primary switch" data-state="unchecked" role="switch" type="button" value="on">
    <span class="slider" data-state="unchecked"></span>
  </button>
</verdocs-switch>
`;
